src/state/selectors/barSelectors.ts

```diff
@@ -0,0 +1,495 @@
+      return { stackId: undefined, dataKeys, barSize };
+    });
+
+    return [...stackedSizeList, ...unstackedSizeList];
+  },
+);
+
+export const selectBarBandSize: (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+  isPanorama: boolean,
+  barSettings: BarSettings,
+) => number | undefined = (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+  isPanorama: boolean,
+  barSettings: BarSettings,
+): number | undefined => {
+  const layout = selectChartLayout(state);
+  const globalMaxBarSize = selectRootMaxBarSize(state);
+  const { maxBarSize: childMaxBarSize } = barSettings;
+  const maxBarSize: number = isNil(childMaxBarSize) ? globalMaxBarSize : childMaxBarSize;
+  let axis: BaseAxisWithScale, ticks: ReadonlyArray<TickItem>;
+  if (layout === 'horizontal') {
+    axis = selectAxisWithScale(state, 'xAxis', xAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'xAxis', xAxisId, isPanorama);
+  } else {
+    axis = selectAxisWithScale(state, 'yAxis', yAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'yAxis', yAxisId, isPanorama);
+  }
+  return getBandSizeOfAxis(axis, ticks, true) ?? maxBarSize ?? 0;
+};
+
+const selectAxisBandSize = (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+  isPanorama: boolean,
+): number | undefined => {
+  const layout = selectChartLayout(state);
+  let axis: BaseAxisWithScale, ticks: ReadonlyArray<TickItem>;
+  if (layout === 'horizontal') {
+    axis = selectAxisWithScale(state, 'xAxis', xAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'xAxis', xAxisId, isPanorama);
+  } else {
+    axis = selectAxisWithScale(state, 'yAxis', yAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'yAxis', yAxisId, isPanorama);
+  }
+  return getBandSizeOfAxis(axis, ticks);
+};
+
+function getBarPositions(
+  barGap: string | number,
+  barCategoryGap: string | number,
+  bandSize: number,
+  sizeList: SizeList,
+  maxBarSize: number,
+): ReadonlyArray<BarWithPosition> {
+  const len = sizeList.length;
+  if (len < 1) return null;
+
+  let realBarGap = getPercentValue(barGap, bandSize, 0, true);
+
+  let result: ReadonlyArray<BarWithPosition>;
+  const initialValue: ReadonlyArray<BarWithPosition> = [];
+
+  // whether is barSize set by user
+  // Okay but why does it check only for the first element? What if the first element is set but others are not?
+  if (sizeList[0].barSize === +sizeList[0].barSize) {
+    let useFull = false;
+    let fullBarSize = bandSize / len;
+    let sum = sizeList.reduce((res, entry) => res + entry.barSize || 0, 0);
+    sum += (len - 1) * realBarGap;
+
+    if (sum >= bandSize) {
+      sum -= (len - 1) * realBarGap;
+      realBarGap = 0;
+    }
+    if (sum >= bandSize && fullBarSize > 0) {
+      useFull = true;
+      fullBarSize *= 0.9;
+      sum = len * fullBarSize;
+    }
+
+    const offset = ((bandSize - sum) / 2) >> 0;
```

Comment:
`>>`? Is a right shift intentional here? If so, why?

Reply:
> Right shifting any number x by 0 returns x converted to a 32-bit integer. Do not use >> 0 to truncate numbers to integers; use [Math.trunc()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/trunc#using_bitwise_no-ops_to_truncate_numbers) instead.

Reply:
Copy paste from the generator.

Reply:
Oof, we should get rid of now imo if we can since it's pretty confusing. But all good, I've never seen it there before

Reply:
Agreed but honestly it's quite low on my list of priorities. We have other things to do first imho.

Reply:
Yep no worries